### PR TITLE
feat(mission_planner): refine goal pose with parameter and add config file

### PR DESCRIPTION
**Pull request by ismetatabay**
## Description

This PR adds config file to mission_planner and  parameter for refining goal pose in order to skip goal validation step. If added parameter is set to true, then given goal is refined according to the closest lane orientation.

autoware_launch PR (must be merged with this PR): https://github.com/autowarefoundation/autoware_launch/pull/225

## Tests performed

https://user-images.githubusercontent.com/56237550/212691970-35c7803d-7609-4d29-886b-e0a08274dc06.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 1]
```diff
@@ -102,6 +102,31 @@ double project_goal_to_map(
   return project.z();
 }
 
+geometry_msgs::msg::Pose get_closest_centerline_pose(
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:105]: Currently Autoware guidelines recommend lower_snake_case.
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/#use-lower-snake-case-for-function-names-required-partially-automated

[the remark this answers is not in the document]
Comment (mehmetdogru) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:105]: I didn't know it has changed. Ok, thanks a lot! @isamu-takagi

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 2]
```diff
@@ -86,3 +86,30 @@ lanelet::ConstLanelet combine_lanelets(const lanelet::ConstLanelets & lanelets)
...
+geometry_msgs::msg::Pose convertBasicPoint3dToPose(
+  const lanelet::BasicPoint3d & point, const double & lane_yaw)
+{
+  // calculate new orientation of goal
+  geometry_msgs::msg::Pose pose;
+  pose.position.x = point.x();
+  pose.position.y = point.y();
+  pose.position.z = point.z();
+
+  pose.orientation = tier4_autoware_utils::createQuaternionFromRPY(0.0, 0.0, lane_yaw);
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:112]: createQauterninoFromYaw can be used.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.hpp [hunk 3]
```diff
@@ -36,6 +36,7 @@ namespace mission_planner::lanelet2
 struct DefaultPlannerParameters
 {
   double goal_angle_threshold_deg;
+  bool correct_goal_pose;
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.hpp:39]: I'm not sure about the best way to name the boolean flag, but the flag name which starts with a verb sounds a bit weird.

According to https://www.samanthaming.com/tidbits/34-better-boolean-variable-names/, enable/disable + verb may be better?

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 4]
```diff
@@ -86,3 +86,30 @@ lanelet::ConstLanelet combine_lanelets(const lanelet::ConstLanelets & lanelets)
   auto combined_lanelet = lanelet::Lanelet(lanelet::InvalId, left_bound, right_bound);
   return std::move(combined_lanelet);
 }
+
+std::vector<geometry_msgs::msg::Point> convertCenterlineToPoints(const lanelet::Lanelet & lanelet)
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:90]: add const to the function

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 5]
```diff
@@ -380,18 +406,22 @@ PlannerPlugin::LaneletRoute DefaultPlanner::plan(const RoutePoints & points)
     const auto local_route_sections = route_handler_.createMapSegments(path_lanelets);
     route_sections = combine_consecutive_route_sections(route_sections, local_route_sections);
   }
-
-  if (!is_goal_valid(points.back(), all_route_lanelets)) {
```
[the remark this answers is not in the document]
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:384]: Thanks for reviewing, at the corners like below image, the goal footprint exceeds lane and it is rejected. I did this way because avoid this situation,  but you're right, this can have dangerous consequences. I updated it. :+1: 

<p align="center">

</p>

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 6]
```diff
@@ -86,3 +86,30 @@ lanelet::ConstLanelet combine_lanelets(const lanelet::ConstLanelets & lanelets)
...
+{
+  std::vector<geometry_msgs::msg::Point> centerline_points;
+  geometry_msgs::msg::Point center_point;
+  for (const auto & point : lanelet.centerline()) {
+    center_point.x = point.basicPoint().x();
+    center_point.y = point.basicPoint().y();
+    center_point.z = point.basicPoint().z();
+    centerline_points.push_back(tier4_autoware_utils::getPoint(center_point));
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:98]: no need to use getPoint function

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 7]
```diff
@@ -102,6 +102,31 @@ double project_goal_to_map(
...
+  closest_lanelet.setCenterline(refined_center_line);
+
+  const double lane_yaw = lanelet::utils::getLaneletAngle(closest_lanelet, point.position);
+
+  const auto nearest_idx =
+    motion_utils::findNearestIndex(convertCenterlineToPoints(closest_lanelet), point.position);
+  const auto nearest_point = closest_lanelet.centerline()[nearest_idx];
+  // update y coordinate according to the asymmetric vehicle margin
+  const double refined_y =
+    nearest_point.y() - (vehicle_info.left_overhang_m - vehicle_info.right_overhang_m) / 2.0;
```
[the remark this answers is not in the document]
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp]: Thanks for correction, I updated it. :+1:

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 8]
```diff
@@ -86,3 +86,30 @@ lanelet::ConstLanelet combine_lanelets(const lanelet::ConstLanelets & lanelets)
...
+  }
+  return centerline_points;
+}
+
+geometry_msgs::msg::Pose convertBasicPoint3dToPose(
+  const lanelet::BasicPoint3d & point, const double & lane_yaw)
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:104]: reference is not required for primitive type (lane_yaw in this case)

File: planning/mission_planner/launch/mission_planner.launch.xml [hunk 9]
```diff
@@ -9,6 +9,7 @@
     <param name="arrival_check_distance" value="1.0"/>
     <param name="arrival_check_duration" value="1.0"/>
     <param name="goal_angle_threshold_deg" value="45.0"/>
+    <param name="correct_goal_pose" value="true"/>
```
Comment (takayuki5168) [inline planning/mission_planner/launch/mission_planner.launch.xml:12]: After this PR, could you make another PR to move these parameters to yaml file so that integrators can easily change the parameters for each vehicle.

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp (hunk 2 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp]: @takayuki5168 thanks for the review! I updated it.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.hpp (hunk 3 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.hpp]: done :+1:

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 10]
```diff
@@ -86,3 +86,30 @@ lanelet::ConstLanelet combine_lanelets(const lanelet::ConstLanelets & lanelets)
   auto combined_lanelet = lanelet::Lanelet(lanelet::InvalId, left_bound, right_bound);
   return std::move(combined_lanelet);
 }
+
+std::vector<geometry_msgs::msg::Point> convertCenterlineToPoints(const lanelet::Lanelet & lanelet)
+{
+  std::vector<geometry_msgs::msg::Point> centerline_points;
+  geometry_msgs::msg::Point center_point;
```
[the remark this answers is not in the document]
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp]: done :+1:

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp (hunk 6 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp]: done :+1:

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp (hunk 8 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp]: done :+1:

File: planning/mission_planner/launch/mission_planner.launch.xml (hunk 9 above)
Comment (ismetatabay) [reply planning/mission_planner/launch/mission_planner.launch.xml]: @mehmetdogru Sure, you can assign me.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 11]
```diff
@@ -102,6 +102,35 @@ double project_goal_to_map(
...
+  const auto refined_center_line = lanelet::utils::generateFineCenterline(closest_lanelet, 1.0);
+  closest_lanelet.setCenterline(refined_center_line);
+
+  const double lane_yaw = lanelet::utils::getLaneletAngle(closest_lanelet, point.position);
+
+  const auto nearest_idx =
+    motion_utils::findNearestIndex(convertCenterlineToPoints(closest_lanelet), point.position);
+  const auto nearest_point = closest_lanelet.centerline()[nearest_idx];
+
+  // update y coordinate according to the asymmetric vehicle margin
```
Comment (mehmetdogru) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:121]: Maybe sounds better?

```suggestion
  // shift nearest point on its local y axis so that vehicle's right and left edges would have approx the same clearance from road border 
```

Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp]: done :+1:

**Comment by ismetatabay** [comment]
@takayuki5168 Hi, I also added config file for mission planner . Could you check it?

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 12]
```diff
@@ -380,8 +411,13 @@ PlannerPlugin::LaneletRoute DefaultPlanner::plan(const RoutePoints & points)
     const auto local_route_sections = route_handler_.createMapSegments(path_lanelets);
     route_sections = combine_consecutive_route_sections(route_sections, local_route_sections);
   }
+  geometry_msgs::msg::Pose goal_pose;
+  goal_pose = points.back();
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:415]:
```suggestion
  auto goal_pose = points.back();
```

or I prefer
```cpp
const auto goal_pose = [&]() {
  if (param_.enable_correct_goal_pose) {
    return get_closest_centerline_pose(road_lanelets_, goal_pose, vehicle_info_);
  }
  return points.back();
}();
```

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 13]
```diff
@@ -102,6 +102,36 @@ double project_goal_to_map(
...
+  const double lane_yaw = lanelet::utils::getLaneletAngle(closest_lanelet, point.position);
+
+  const auto nearest_idx =
+    motion_utils::findNearestIndex(convertCenterlineToPoints(closest_lanelet), point.position);
+  const auto nearest_point = closest_lanelet.centerline()[nearest_idx];
+
+  // shift nearest point on its local y axis so that vehicle's right and left edges
+  // would have approx the same clearance from road border
+  const auto shift_length = (vehicle_info.right_overhang_m - vehicle_info.left_overhang_m) / 2.0;
+  const auto delta_x = -1 * shift_length * sin(lane_yaw);
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:124]:
```suggestion
  const auto delta_x = -shift_length * std::sin(lane_yaw);
```

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 14]
```diff
@@ -102,6 +102,36 @@ double project_goal_to_map(
...
+  const auto delta_y = shift_length * cos(lane_yaw);
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:125]:
```suggestion
  const auto delta_y = shift_length * std::cos(lane_yaw);
```

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 15]
```diff
@@ -102,6 +102,36 @@ double project_goal_to_map(
...
+
+  lanelet::BasicPoint3d refined_point(
+    nearest_point.x() + delta_x, nearest_point.y() + delta_y, nearest_point.z());
+
+  auto closest_pose = convertBasicPoint3dToPose(refined_point, lane_yaw);
+
+  return closest_pose;
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:132]:
```suggestion
  return convertBasicPoint3dToPose(refined_point, lane_yaw);
```
or 
```cpp
  const auto closest_pose = convertBasicPoint3dToPose(refined_point, lane_yaw);
  return closest_pose;
```

**Review by takayuki5168 — APPROVED** [review]
@ismetatabay Thank you for the fix. I left some additional comments but give you approval beforehand.

**Comment by ismetatabay** [comment]
@takayuki5168 thanks for the review, i updated the branch with your suggestions :+1: .

**Comment by takayuki5168** [comment]
LGTM

**Review by mehmetdogru — APPROVED** [review]
